File: Algorithms/tree/102.binary-tree-level-order-traversal/binary-tree-level-order-traversal.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# 
# Copyright (c) 2017 gosyang. All Rights Reserved
# 
"""
File: binary-tree-level-order-traversal
Author: gosyang
Date: 2020/2/4 5:12 PM
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        deq, res = deque([0, root],), []
        while deq:
            r = deq.popleft()

File: Algorithms/tree/102.binary-tree-level-order-traversal/binary-tree-level-order-traversal.go
// Package _02_binary_tree_level_order_traversal - package brief introduce
/* Copyright 2019 All Rights Reserved. */
/* binary-tree-level-order-traversal - file brief introduce */
/*
modification history
-----------------------------
2020/4/18 4:39 PM, by gosyang, create
*/
/*
给你一个二叉树，请你返回其按 层序遍历 得到的节点值。 （即逐层地，从左到右访问所有节点）。
*/
package _02_binary_tree_level_order_traversal

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	res := make([][]int, 0)
	if root == nil {
		return res
	}
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		size := len(queue)
		tmp := []int{}

		for i := 0; i < size; i++ {
			tmp = append(tmp, queue[i].Val)
			if queue[i].Left != nil {
				queue = append(queue, queue[i].Left)
			}
			if queue[i].Right != nil {
				queue = append(queue, queue[i].Right)
			}
		}
		queue = queue[size:]
		res = append(res, tmp)
	}
	return res
}

// 最简单的如果结果是一维数组，就直接每次tmp的地方res append就行。关键是这里是批量bfs用熟